README.md
# urbit-ob

[![Build Status](https://secure.travis-ci.org/urbit/urbit-ob.png)](http://travis-ci.org/urbit/urbit-ob)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![npm](https://img.shields.io/npm/v/urbit-ob.svg)](https://www.npmjs.com/package/urbit-ob)

Utilities for phonetic base wrangling.

## What

Here you can primarily find functions for dealing with the *phonetic bases*
used by Urbit.  The `@p` encoding is used for naming ships, while the `@q`
encoding is used for representing arbitrary data in a memorable and
pronounceable fashion.

The `@p` encoding is an *obfuscated* representation of an underlying 32-bit
number, in particular, hence the 'ob' in the library's name.

## Install

You can grab the library from npm via:

```
npm install urbit-ob
```

Otherwise you can just clone the repo, and a simple `npm install` will pull
down the dependencies.

## Usage

The library exposes two families of functions:

* `patp / patp2dec / patp2hex / hex2patp / isValidPatp`
* `patq / patq2dec / patq2hex / hex2patq / isValidPatq`

They are pretty self-explanatory.  Use `patp` or `patq` to convert base-10
numbers (or strings encoding base-10 numbers) to `@p` or `@q` respectively.
Use `patp2dec` or `patq2dec` to go in reverse.  `patp2hex`, `patq2hex`, and
their inverses work similarly.  `isValidPat{p, q}` can be used to check the
validity of a `@p` or `@q` string.

Some examples:

```
> const ob = require('urbit-ob')
> ob.patp('0')
'~zod'
> ob.patp2dec('~nidsut-tomdun')
'15663360'
> ob.hex2patq('010203')
'~doznec-binwes'
> ob.patq2hex('~marned-wismul-nilsev-botnyt')
'01ca0e51d20462f3'
> ob.isValidPatq('~marned-wismul-nilsev-botnyt')
> true
```

There are a few other noteworthy functions exposed as well:

* `clan`, for determining the ship class of a `@p` value
* `sein`, for determining the parent of a `@p` value
* `eqPatq`, for comparing `@q` values for equality
* `isValidPat`, for a faster/weaker check of `@p` or `@q`-ness that only
  validates syllables (and not proper dash formatting)

For example, you can check that `~marzod` is a star with parent `~zod`:

```
> ob.clan('~marzod')
'star'
> ob.sein('~marzod')
'~zod'
```

And note that `@q` values are considered equal modulo the existence of leading
zero bytes:

```
> '~doznec-marzod' === '~nec-marzod'
false
> ob.eqPatq('~doznec-marzod', '~nec-marzod')
true
```

## Testing

A simple `npm test` will run the test suite.


## Building

`npm run build` will build the library for the browser.

## See Also

* [urbit-hob](https://github.com/urbit/urbit-hob) -- Haskell bindings


CHANGELOG
# Changelog

- 5.0.1 (2020-06-26)
    * Fixes a bug where some functions (namely, 'patp') could enter an
      infinite loop when supplied with a null value.

- 5.0.0 (2020-03-24)
    * The isValidPat{p, q} functions now perform stricter validations on their
      inputs.
    * Adds an 'isValidPat' function that implements the old behaviour of
      isValidPatp/isValidPatq for faster, but weaker, string validation.

- 4.1.4 (2019-09-09)
    * Further fixes to the 'browser' specification [1].  Drops the single-line
      "simple" spec in favour of the object version.  In this case we give a
      hint to downstream bundlers to replace instances of "urbit-ob" with the
      browser bundle we distribute.

- 4.1.3 (2019-09-09)
    * Further fixes to the 'browser' specification.  Uses the single-line
      spec outlined in [1] that simply provides an entry point hint for
      builders.

      [1]: https://github.com/defunctzombie/package-browser-field-spec

- 4.1.2 (2019-09-05)
    * Reverts change that likely caused the 'browser' specification in
      package.json to break.

- 4.1.1 (2019-08-07)
    * Reverts change that messed up the library's namespacing downstream.

- 4.1.0 (2019-08-07)
    * Adds a browser build.  The library should now "just work" when imported
      in a browser app, instead of requiring building downstream. (#24)

- 4.0.1 (2019-06-17)
    * Fixes a bug in isValidPat{p, q} that caused it to return 'true' for '~'.
      (#23)

- 4.0.0 (2019-03-18)
    * Fixes the @p collision bug described in urbit/arvo#1105.

      An error in the Hoon implementation of a permutation cipher used to
      encipher planet-sized atoms resulted in the cipher not being bijective,
      occasionally causing different inputs to produce the same output.

      This version updates the 'patp' family of functions to support the
      corrected cipher.

- 3.2.0 (2019-02-08)
    * Improvements to error reporting.  E.g., instead of 'clan("~nidso")'
      throwing an error labelled 'patp2hex', it will now throw one labelled
      'clan'.

- 3.1.1 (2018-12-01)
    * Short-circuit on isValidPatp and isValidPatq when there is no leading
      tilde.

- 3.1.0 (2018-11-16)
    * Exports 'isValidPatp' and 'isValidPatq' functions for checking the
      validity of @p and @q strings.

- 3.0.0 (2018-11-02)
    * Major library cleanup and API simplification.  Eliminates old, dead,
      or redundant code that had accrued over time.  Exposes patp, patp2dec,
      patp2hex, and hex2patp functions as the API core (and similarly for @q).
    * Adds proper in-library documentation, and beefs up the README.
    * Adds property tests for both internals and exposed functions.
    * Standardises output, e.g. by always including a leading tilde in @p
      values.

- 2.0.0 (2018-10-22)
    * Bundle/distribute as CommonJS.

- 1.4.0 (2018-10-10)
    * Add 'eqPatq' function for comparing @q values for equality (i.e., modulo
      leading zero bytes).


test/med.js
const BN = require('bn.js')
const { expect } = require('chai')
const isEqual = require('lodash.isequal')

const { Fe, Fen } = require('../src/internal/ob')

const u_a = new BN(Math.pow(2, 4) - 1)
const u_b = new BN(Math.pow(2, 4))
const u_c = u_a.mul(u_b)

const emm = [
  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,  14,  15,
  16,  17, 18,  19,  20,  21,  22,  23,  24,  25,  26,  27,  28,  29,  30,  31,
  32,  33,  34,  35, 36,  37,  38,  39,  40,  41,  42,  43,  44,  45,  46,  47,
  48,  49,  50,  51,  52,  53, 54,  55,  56,  57,  58,  59,  60,  61,  62,  63,
  64,  65,  66,  67,  68,  69,  70,  71, 72,  73,  74,  75,  76,  77,  78,  79,
  80,  81,  82,  83,  84,  85,  86,  87,  88,  89, 90,  91,  92,  93,  94,  95,
  96,  97,  98,  99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111,
  112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127,
  128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
  144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159,
  160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175,
  176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191,
  192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207,
  208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223,
  224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239
  ]

const v0 = [
  106,  54,  57, 110, 216, 157,  90, 138, 148, 205, 214, 229,  25, 104, 217,  70,  16,  91,
  180, 108, 189, 176,  67, 213, 154, 194, 122, 199, 136, 140,  36,  56,  87, 112,   8,  34,
   14, 171, 227, 160, 211, 228,  37, 121, 119,  65, 132,  45, 224,  61, 141,  59,  82,  77,
   74,  20, 130, 181, 123, 186, 166,  42,  81, 172, 105, 196,  44, 135, 156, 192, 116,  39,
    7,  40,  84, 169, 193, 131,  88, 142,  24, 128,  38, 222, 197, 218, 159,  30, 145,  58,
   53,  85,  62,  49, 158,  86,  72, 210, 225,  52,  73, 149, 143, 195, 124, 179, 219,   9,
  200,  64,  51,  48,  26, 234,  27, 232, 231, 153, 190, 133, 109, 126,   6, 178, 183, 151,
  117,  46, 161,  43, 185, 236, 127,  89, 223,  23,  69,  68, 209, 139,  19,  33,  79, 164,
  207,  50, 144,  31, 134, 170,  29, 107, 220, 184,  47, 103, 206, 201, 175, 125,  35, 114,
  146,  10,  55, 152,  98,   1, 168, 215,  28, 237, 101,  17, 155, 118,  83, 147, 115, 100,
  233,   4,  66,   0, 150, 203,  22,   5, 174,  11,  18, 177,   3, 165,  99, 167, 202, 212,
  163, 182,  80, 162,  71,  97,  12,  60, 113, 221, 204,  41, 226, 187,  63, 230,   2, 188,
  208,  76, 191, 235,  93,  13, 111, 238,  78, 198,  21,  92,  95,  94,  96, 102, 120, 239,
  32,  129,  15, 173, 137,  75
  ]

const eff = (_, m) => new BN(v0[m])

const feis = arg =>
  Fe(4, u_a, u_b, u_c, eff, new BN(arg))

const tail = arg =>
  Fen(4, u_a, u_b, u_c, eff, new BN(arg))

// test

describe('feis/tail (medium input space)', () => {

  const perm = emm.map(x => feis(x).toNumber())
  const inv  = perm.map(x => tail(x).toNumber())
  const distincts = perm.filter((x, i, a) => a.indexOf(x) === i)

  it('feis produces distinct outputs', () => {
    expect(distincts.length).to.equal(perm.length)
  })

  it('feis permutes the input space', () => {
    expect(perm.reduce((acc, x) => emm.includes(x) && acc, true))
    .to.equal(true)
  })

  it('tail inverts feis', () => {
    expect(isEqual(emm, inv)).to.equal(true)
  })

})


test/main.js
//var assert = require('assert');
var nom = require('../index.js');
//var bn = require('bignum');

exports['test that stops execution'] = function(assert) {
  assert.equal(nom._feen(0x10100), 0x63b30e1c, 'test feen');
  assert.equal(nom._fend(0x63b30e1c), 0x10100, 'test fend');
  assert.equal(nom._muk(0, 2, 0x101), 0x42081a9b, 'test muk 1');
  assert.equal(nom._muk(0, 2, 0x201), 0x64c7667e, 'test muk 2');

  // assert.equal(nom._wordtonum('zod'), 0, 'test converting one syllable word to num');
  // assert.equal(nom._wordtonum('samzod'), 1024, 'test converting two syllable word to num');
  // assert.equal((typeof nom._wordtonum('sambinzod')), 'undefined', 'test rejecting three syllable word to num');

  assert.equal(nom.toAddress('lex'), 200, 'convert ship name ~lex to number 200');
  assert.equal(nom.toAddress('binzod'), 512, 'convert ship name ~binzod to number 512');
  assert.equal(nom.toAddress('poldec-tonteg'), 9896704, 'convert ship name ~poldec-tonteg to number 9896704');
  assert.equal(nom._getsuffix(200), 'lex', 'gets correct suffix');
  assert.equal(nom.toGalaxyName(200), 'lex', 'convert ship num 200 to ~lex');
  assert.equal(nom.toStarName(512), 'binzod', 'convert ship num 512 to ~binzod');
  assert.equal(nom.toPlanetName(9896704), 'poldec-tonteg', 'convert number 9896704 to ship name ~poldec-tonteg');
  assert.equal(nom.isValidName('poldec-tonteg'), true, 'identifies poldec-tonteg as a ship');
  assert.equal(nom.isValidName('zod'), true, 'identifies zod as a ship');
  assert.equal(nom.isValidName('aaa'), false, 'identifies aaa as not ship');
  assert.equal(nom.isValidName("aaaaaa"), false, 'invalidates star');
  assert.equal(nom.isValidName("//////"), false, 'invalidates star');
  assert.equal(nom.isValidName("////////////"), false, 'invalidates planet');
  assert.equal(nom.isValidName("//////-//////"), false, 'invalidates planet');
}

if (module == require.main) require('test').run(exports);


test/test.js
// //var assert = require('assert');
var ob = require('../src/index.js');
// //var bn = require('bignum');

// import {_feen} from '../src/index'



test('test feen', () => {
  expect(ob._feen(0x10100)).toEqual(0x63b30e1c);
});

test('test fend', () => {
  expect(ob._fend(0x63b30e1c)).toEqual(0x10100);
});

test('test muk 1', () => {
  expect(ob._muk(0, 2, 0x101)).toEqual(0x42081a9b);
});

test('test muk 2', () => {
  expect(ob._muk(0, 2, 0x201)).toEqual(0x64c7667e);
});

test('test toAddress lex', () => {
  expect(ob.toAddress('lex')).toEqual('200');
});

test('test toAddress samzod', () => {
  expect(ob.toAddress('samzod')).toEqual('1024');
});

test('test toAddress sambinzod', () => {
  expect(ob.toAddress('sambinzod')).toEqual(undefined);
});

test('test toAddress poldec-tonteg', () => {
  expect(ob.toAddress('poldec-tonteg')).toEqual(9896704);
});

// exports['test that stops execution'] = function(assert) {
//   assert.equal(nom._feen(0x10100), 0x63b30e1c, 'test feen');
//   assert.equal(nom._fend(0x63b30e1c), 0x10100, 'test fend');
//   assert.equal(nom._muk(0, 2, 0x101), 0x42081a9b, 'test muk 1');
//   assert.equal(nom._muk(0, 2, 0x201), 0x64c7667e, 'test muk 2');
//
//   assert.equal(nom._wordtonum('zod'), 0, 'test converting one syllable word to num');
//   assert.equal(nom._wordtonum('samzod'), 1024, 'test converting two syllable word to num');
//   assert.equal((typeof nom._wordtonum('sambinzod')), 'undefined', 'test rejecting three syllable word to num');
//
//
//   assert.equal(nom.toAddress('lex'), 200, 'convert ship name ~lex to number 200');
//   assert.equal(nom.toAddress('binzod'), 512, 'convert ship name ~binzod to number 512');
//   assert.equal(nom.toAddress('poldec-tonteg'), 9896704, 'convert ship name ~poldec-tonteg to number 9896704');
//   assert.equal(nom._getsuffix(200), 'lex', 'gets correct suffix');
//   assert.equal(nom.toGalaxyName(200), 'lex', 'convert ship num 200 to ~lex');
//   assert.equal(nom.toStarName(512), 'binzod', 'convert ship num 512 to ~binzod');
//   assert.equal(nom.toPlanetName(9896704), 'poldec-tonteg', 'convert number 9896704 to ship name ~poldec-tonteg');
//   assert.equal(nom.isShip('poldec-tonteg'), true, 'identifies poldec-tonteg as a ship');
//   assert.equal(nom.isShip('aaa'), false, 'identifies aaa as not ship');
//   assert.equal(nom.isShip("aaaaaa"), false, 'invalidates star');
//   assert.equal(nom.isShip("//////"), false, 'invalidates star');
//   assert.equal(nom.isShip("////////////"), false, 'invalidates planet');
//   assert.equal(nom.isShip("//////-//////"), false, 'invalidates planet');
// }

// if (module == require.main) require('test').run(exports);


gulpfile.js
var gulp = require('gulp');
var rollup = require('gulp-better-rollup');
var sucrase = require('@sucrase/gulp-plugin');
var resolve = require('rollup-plugin-node-resolve');
var commonjs = require('rollup-plugin-commonjs');
var replace = require('rollup-plugin-replace');
var builtins = require('rollup-plugin-node-builtins');
var globals = require('rollup-plugin-node-globals');
var rootImport = require('rollup-plugin-root-import');

gulp.task('default', function(cb) {
  return gulp
    .src('src/index.js')
    .pipe(rollup({
      plugins: [
        // the order of these plugins matter
        commonjs(),
        replace({
          'process.env.NODE_ENV': JSON.stringify('development')
        }),
        rootImport({
          root: `${__dirname}/dist`,
          useEntry: 'prepend',
          extensions: '.js'
        }),
        globals(),
        builtins(),
        resolve()
      ]
    }, 'cjs'))
    .on('error', function(e){
      console.log(e);
      cb();
    })
    .pipe(gulp.dest('./dist/'))
    .on('end', cb);
});


src/internal/ob.js
// ++  ob
//
// See arvo/sys/hoon.hoon.

const BN = require('bn.js')
const { muk } = require('./muk')

const ux_1_0000 = new BN('10000', 'hex')
const ux_ffff_ffff = new BN('ffffffff', 'hex')
const ux_1_0000_0000 = new BN('100000000', 'hex')
const ux_ffff_ffff_ffff_ffff = new BN('ffffffffffffffff', 'hex')
const ux_ffff_ffff_0000_0000 = new BN('ffffffff00000000', 'hex')

const u_65535 = new BN('65535')
const u_65536 = new BN('65536')

// a PRF for j in { 0, .., 3 }
const F = (j, arg) => {
  const raku = [
    0xb76d5eed,
    0xee281300,
    0x85bcae01,
    0x4b387af7,
  ]

  return muk(raku[j], 2, arg)
}

/**
 * Conceal structure v3.
 *
 * @param {String, Number, BN} pyn
 * @return  {BN}
 */
const fein = (arg) => {
  const loop = (pyn) => {
    const lo = pyn.and(ux_ffff_ffff)
    const hi = pyn.and(ux_ffff_ffff_0000_0000)

    return pyn.gte(ux_1_0000) && pyn.lte(ux_ffff_ffff)
      ? ux_1_0000.add(feis(pyn.sub(ux_1_0000)))
      : pyn.gte(ux_1_0000_0000) && pyn.lte(ux_ffff_ffff_ffff_ffff)
      ? hi.or(loop(lo))
      : pyn
  }

  return loop(new BN(arg))
}

/**
 * Restore structure v3.
 *
 * @param  {String, Number, BN}  cry
 * @return  {BN}
 */
const fynd = (arg) => {
  const loop = (cry) => {
    const lo = cry.and(ux_ffff_ffff)
    const hi = cry.and(ux_ffff_ffff_0000_0000)

    return cry.gte(ux_1_0000) && cry.lte(ux_ffff_ffff)
      ? ux_1_0000.add(tail(cry.sub(ux_1_0000)))
      : cry.gte(ux_1_0000_0000) && cry.lte(ux_ffff_ffff_ffff_ffff)
      ? hi.or(loop(lo))
      : cry
  }

  return loop(new BN(arg))
}

/**
 * Generalised Feistel cipher.
 *
 * See: Black and Rogaway (2002), "Ciphers with arbitrary finite domains."
 *
 * Note that this has been adjusted from the reference paper in order to
 * support some legacy behaviour.
 *
 * @param  {String, Number, BN}
 * @return  {BN}
 */
const feis = arg =>
  Fe(4, u_65535, u_65536, ux_ffff_ffff, F, new BN(arg))

const Fe = (r, a, b, k, f, m) => {
  const c = fe(r, a, b, f, m)
  return (
      c.lt(k)
    ? c
    : fe(r, a, b, f, c)
  )
}

const fe = (r, a, b, f, m) => {
  const loop = (j, ell, arr) => {
    if (j > r) {
      return (
          r % 2 !== 0
        ? a.mul(arr).add(ell)
        : arr.eq(a)
        ? a.mul(arr).add(ell)
        : a.mul(ell).add(arr)
      )
    } else {
      const eff = f(j - 1, arr)

      const tmp =
          j % 2 !== 0
        ? ell.add(eff).mod(a)
        : ell.add(eff).mod(b)

      return loop(j + 1, arr, tmp)
    }
  }

  const L = m.mod(a)
  const R = m.div(a)

  return loop(1, L, R)
}

/**
 * Reverse 'feis'.
 *
 * See: Black and Rogaway (2002), "Ciphers with arbitrary finite domains."
 *
 * Note that this has been adjusted from the reference paper in order to
 * support some legacy behaviour.
 *
 * @param {Number, String, BN}  arg
 * @return  {BN}
 */
const tail = arg =>
  Fen(4, u_65535, u_65536, ux_ffff_ffff, F, new BN(arg))

const Fen = (r, a, b, k, f, m) => {
  const c = fen(r, a, b, f, m)
  return (
      c.lt(k)
    ? c
    : fen(r, a, b, f, c)
  )
}

const fen = (r, a, b, f, m) => {
  const loop = (j, ell, arr) => {
    if (j < 1) {
      return a.mul(arr).add(ell)
    } else {
      const eff = f(j - 1, ell)

      // NB (jtobin):
      //
      // Slight deviation from B&R (2002) here to prevent negative values.  We
      // add 'a' or 'b' to arr as appropriate and reduce 'eff' modulo the same
      // number before performing subtraction.
      //
      const tmp =
          j % 2 !== 0
        ? arr.add(a).sub(eff.mod(a)).mod(a)
        : arr.add(b).sub(eff.mod(b)).mod(b)

      return loop(j - 1, tmp, ell)
    }
  }

  const ahh =
      r % 2 !== 0
    ? m.div(a)
    : m.mod(a)

  const ale =
      r % 2 !== 0
    ? m.mod(a)
    : m.div(a)

  const L =
      ale.eq(a)
    ? ahh
    : ale

  const R =
      ale.eq(a)
    ? ale
    : ahh

  return loop(r, L, R)
}

module.exports = {
  F,

  fe,
  Fe,
  feis,
  fein,

  fen,
  Fen,
  tail,
  fynd
}


nom.js
'use strict';

/*
 * Utility methods
 */

//var bn = require('bignum');
var bnjs = require('bn.js');

var raku = [
  3077398253,
  3995603712,
  2243735041,
  1261992695
];

var prefix = "dozmarbinwansamlitsighidfidlissogdirwacsabwissibrigsoldopmodfoglidhopdardorlorhodfolrintogsilmirholpaslacrovlivdalsatlibtabhanticpidtorbolfosdotlosdilforpilramtirwintadbicdifrocwidbisdasmidloprilnardapmolsanlocnovsitnidtipsicropwitnatpanminritpodmottamtolsavposnapnopsomfinfonbanmorworsipronnorbotwicsocwatdolmagpicdavbidbaltimtasmalligsivtagpadsaldivdactansidfabtarmonranniswolmispallasdismaprabtobrollatlonnodnavfignomnibpagsopralbilhaddocridmocpacravripfaltodtiltinhapmicfanpattaclabmogsimsonpinlomrictapfirhasbosbatpochactidhavsaplindibhosdabbitbarracparloddosbortochilmactomdigfilfasmithobharmighinradmashalraglagfadtopmophabnilnosmilfopfamdatnoldinhatnacrisfotribhocnimlarfitwalrapsarnalmoslandondanladdovrivbacpollaptalpitnambonrostonfodponsovnocsorlavmatmipfip"

var suffix = "zodnecbudwessevpersutletfulpensytdurwepserwylsunrypsyxdyrnuphebpeglupdepdysputlughecryttyvsydnexlunmeplutseppesdelsulpedtemledtulmetwenbynhexfebpyldulhetmevruttylwydtepbesdexsefwycburderneppurrysrebdennutsubpetrulsynregtydsupsemwynrecmegnetsecmulnymtevwebsummutnyxrextebfushepbenmuswyxsymselrucdecwexsyrwetdylmynmesdetbetbeltuxtugmyrpelsyptermebsetdutdegtexsurfeltudnuxruxrenwytnubmedlytdusnebrumtynseglyxpunresredfunrevrefmectedrusbexlebduxrynnumpyxrygryxfeptyrtustyclegnemfermertenlusnussyltecmexpubrymtucfyllepdebbermughuttunbylsudpemdevlurdefbusbeprunmelpexdytbyttyplevmylwedducfurfexnulluclennerlexrupnedlecrydlydfenwelnydhusrelrudneshesfetdesretdunlernyrsebhulrylludremlysfynwerrycsugnysnyllyndyndemluxfedsedbecmunlyrtesmudnytbyrsenwegfyrmurtelreptegpecnelnevfes"

var getsyllable = function(s, i) {
  return s.slice(i * 3, (i * 3) + 3);
};

var getprefix = function(i) {
  return getsyllable(prefix, i);
};

var getsuffix = function(i) {
  return getsyllable(suffix, i);
};

var getsyllableindex = function(str, syl) {
  var i = str.indexOf(syl);
  if (i < 0) {
    return;
  };
  return i / 3
};

var getprefixindex = function(syl) {
  return getsyllableindex(prefix, syl);
};
  
var getsuffixindex = function(syl) {
  return getsyllableindex(suffix, syl);
};

var wordtonum = function(word) {
    if (word.length == 3) {
      return 1 * getsuffixindex(word);
    } else if (word.length == 6) {
      var addr = getprefixindex(word.slice(0, 2));
      addr = addr * 0x100;
      addr = addr + getsuffixindex(word.slice(3, 5));
      return addr;
    } else {
      return;
    }
};

var feen = function(pyn) {
  var f = 4294967295
  if (pyn >= 0x10000 && pyn <= 0xFFFFFFFF) {
    var tmp = fice(pyn - 0x10000) + 0x10000;
    return tmp;
  }
  if (pyn >= 0x100000000 && pyn <= 0xffffffffffffffff) {
    var pynBn = new bnjs(pyn);
    var lo = pynBn.and(f);
    var hi = pynBn.and('18446744069414584000');
    return hi.or(feen(lo)).toNumber();
  }
  return pyn;
}

var fend = function(cry) {
  if (cry >= 0x10000 && cry <= 0xFFFFFFFF) {
    var res = new bnjs(teil(cry - 0x10000));
    res = res.add(new bnjs(65536)).toNumber();
    return res;
  };
  if (cry >= 0x100000000 && cry <= bn(0xffffffffffffffff)) {
    var cryBn = new bnjs(cry);
    var lo = cryBn.and(new bnjs('0xFFFFFFFF'));
    var hi = cryBn.and(new bnjs('0xffffffff00000000'));
    var res = hi.or(fend(lo));
    return res.toNumber();
  };
  return cry;
};

var fice = function(nor) {
  var sel = [
    nor % 65535,
    nor / 65535
  ];
  for (var i = 0; i < 4; i++) {
    sel = rynd(i, sel[0], sel[1]);
  };

  var res = 65535 * sel[0] + sel[1];
  return res;
};

var teil = function(vip) {
  var sel = [
    vip % 65535,
    vip / 65535
    //vip % 0xFFFF,
    //vip / 0x10000
  ];
  // maybe the for loops got borked in lua conversion
  for (var i = 3; i > -1; i--) {
    sel = rund(i, sel[0], sel[1]);
  };
  //var res = bn(bn(0xFFFF).mul(sel[0])).add(sel[1]);
  var r1 = new bnjs(65535);
  var res = r1.mul(new bnjs(sel[0])).add(new bnjs(sel[1]));
  return res.toNumber();
};

var rynd = function(n, l, r) {
  l = Math.floor(l);
  var res = [r, 0];
  var m = new bnjs(65536);
  if (n % 2 == 0) {
    m = new bnjs(65535);
  };
  //res[1] = (bn(muk(raku[n], 2, r)).add(l)) % m;
  var r1 = new bnjs(muk(raku[n], 2, r));
  var r2 = r1.add(new bnjs(l)).mod(m);
  res[1] = r2.toNumber();
  return res
};

var rund = function(n, l, r) {
  l = Math.floor(l);
  var res = [r, 0];
  var m = new bnjs(65536);
  if (n % 2 == 0) {
    m = new bnjs(65535);
  };
  var h = new bnjs(muk(raku[n], 2, r));
  var r1 = new bnjs(m + l);
  var r2 = r1.sub(h.mod(m)).mod(m).toString();
  res[1] = r2;
  return res
};

var muk = function(syd, len, key) {
  //key = bn(key);
  var lo = key & 0xFF;
  var hi = (key & 0xFF00) / 256;
  var res = murmur3(String.fromCharCode(lo) + String.fromCharCode(hi), syd);
  return res;
};

var murmur3 = function(data, seed) {
  if (!seed) {
    seed = 0;
  }
  var c1 = new bnjs(3432918353);
  var c2 = new bnjs(461845907);

  var f = 4294967295

  var length = new bnjs(data.length);
  var h1 = new bnjs(seed);
  var k1;
  var roundedEnd = length & 0xFFFFFFFC;
  // this will likely need to be redone with bignum
  for (var i = 0; i < roundedEnd; i += 4) {
    var x = data.charCodeAt(i + 3) ? data.charCodeAt(i + 3) : 0;
    k1 = bn(data.charCodeAt(i) & 0xFF)
      | ((data.charCodeAt(i + 1) & 0xFF) << 8)
      | ((data.charCodeAt(i + 2) & 0xFF) << 16)
      | (x << 24);
    k1 = k1 * c1;
    k1 = (k1 << 15) | ((k1 & 0xFFFFFFFF) >> 17);
    k1 = k1 * c2;
    h1 = h1 ^ k1;
    h1 = (h1 << 13) | ((h1 & 0xFFFFFFFF) >> 19);
    h1 = h1 * 5 + 3864292196;
  };

  k1 = 0;
  var val = length & 0x03;
  if (val == 3) {
    k1 = (data.charCodeAt(roundedEnd + 2) & 0xFF) << 16;
  };
  if (val == 3 || val == 2) {
    k1 = k1 | (data.charCodeAt(roundedEnd + 1) & 0xFF) << 8;
  };
  if (val == 3 || val == 2 || val == 1) {
    k1 = k1 | (data.charCodeAt(roundedEnd) & 0xFF);
    k1 = new bnjs(k1 * c1);
    var k2 = new bnjs(k1.and(new bnjs(f)).shrn(17));
    k1 = k1.shln(15).or(k2);
    k1 = k1.mul(c2);
    h1 = h1.xor(k1);
  };
  h1 = h1.xor(length);
  h1 = h1.xor(h1.and(new bnjs(f)).shrn(16));
  h1 = h1.mul(new bnjs(2246822507));
  h1 = h1.xor(h1.and(new bnjs(f)).shrn(13));
  h1 = h1.mul(new bnjs(3266489909));
  h1 = h1.xor(h1.and(new bnjs(f)).shrn(16));
  return h1.and(new bnjs(f)).toNumber();
};

/*
 * Public methods
 *  -- toAddress ( ship name )
 *  --   => address number
 *  -- to{Galaxy,Star,Planet}Ship ( address number )
 *  --   => ship name
 *  ####################### NOT YET IMPLEMENTED ##########################
 *  -- clan ( ship name or address number (int or bn) )
 *  --   => "galaxy", "star", "planet", "moon" or "comet"
 *  -- sein ( ship name or address number (int or bn) )
 *  --   => parent name or address number
 *
 */

var toAddress = function(name, unscramble) {
  if (!unscramble) {
    unscramble = true;
  };
  // our method of planet discovery needs hep
  name = name.replace(/~/g, '');
  if (name.length == 3) {
    return getsuffixindex(name);
  } else if (name.length == 6) {
    var addr = getprefixindex(name.slice(0, 3));
    addr = addr * 256;
    addr = addr + getsuffixindex(name.slice(3, name.length));
    return addr;
  } else if (name.length == 13) {
    var addr = toAddress(name.slice(0, 6));
    addr = addr * 65536;
    addr = addr + toAddress(name.slice(7, addr.length));
    if (unscramble) {
      addr = fend(addr);
    };
    return addr;
  } else {
    return;
  }
};


//var toAddress = function(name) {
//    var nome = name.replace(/~|-/g, '');
//    var lent = nome.length;
//
//    if (lent % 3 != 0) {
//      return;
//    }
//
//    var syls = lent / 3;
//    if (syls > 1 && syls % 2 != 0) {
//      return;
//    }
//
//    if (syls == 1) {
//      return bn(wordtonum(nome))
//    } else if (syls >= 4 && syls <= 8) {
//      var padr = wordtonum(nome.slice(lent - 12, lent - 6));
//      padr = padr * 256
//      padr = padr + wordtonum(nome.slice(lent - 6, lent));
//      padr = fend(padr);
//
//      if (syls == 4) {
//        return padr;
//      };
//
//      var addr = 0;
//      for (var i = 0; i <= syls - 6; i += 2) {
//        addr = addr + wordtonum(nome.slice(i * 3, i * 3 + 6));
//        addr = addr * 0x10000;
//      };
//      return (addr * 0x10000) + padr;
//    } else {
//      var addr = 0;
//
//      for (var i = 0; i <= syls - 2; i +=2) {
//        addr = addr * 0x10000;
//        addr = addr + wordtonum(nome.slice(i * 3, i * 3 + 6));
//      };
//      return addr;
//    }
//
//}

var toGalaxyName = function(galaxy) {
  return toShipName(galaxy, 1);
};

var toStarName = function(star) {
  return toShipName(star, 2);
};

// better ES6 better way to do this
var toPlanetName = function(scrambled, scramble) {
  if (!scramble) {
    scramble = true;
  }
  return toShipName(scrambled, 4, scramble);
};

var toShipName = function(addr, minBytes, scramble) {
  if (!scramble) {
    scramble = true;
  };

  if (!minBytes) {
    if (addr < 0x100) {
      minBytes = 1;
    } else if (addr < 0x10000) {
      minBytes = 2;
    } else {
      minBytes = 4;
    }
  }

  if (minBytes == 4 && scramble) {
    addr = feen(addr);
  };

  var name = ""
  for (var i = 0; i < minBytes; i ++) {
    var byt = Math.floor(addr % 256);
    var syllable = "";
    if (i % 2 == 1) {
      syllable = getprefix(byt);
    } else {
      syllable = getsuffix(byt);
    }
    if (i == 2) {
      name = "-" + name
    };
    name = syllable + name;
    addr = addr / 256;
  };
  return name;
};

module.exports = {
  toAddress: toAddress,
  toGalaxyName: toGalaxyName,
  toStarName: toStarName,
  toPlanetName: toPlanetName,
  _wordtonum: wordtonum,
  _getsuffix: getsuffix,
  _muk: muk,
  _feen: feen,
  _fend: fend,
  _teil: teil
};
